### PR TITLE
Allow subHeaders to be a callback

### PR DESCRIPTION
**Pull request by vigie**
This is the fix to issue https://github.com/stomp-js/rx-stomp/issues/298

Without this, a stale value can be passed through a header during an
automatic resubscribe

I am moving the version with this MR to 1.1.0 as this should be considered
a non-breaking feature addition.

### CODE REVIEW

**Comment by vigie** [comment]
A note on testing: 

I tested this against my own enterprise app.

Before this change, my broker would send error frames to the client following an automatic resubscription that followed a reconnect, if the previous auth-token had expired.

With this change in place I can see the new header values being sent during resubscription and the issue is fixed.

I saw the relevant new lines contributed by this MR executing at the correct subscription and resubscription times using Chrome debugger.

**Comment by kum-deepak** [comment]
Thanks, it makes sense. I have the following suggestions:

- Allow the callback to be optionally async - this will facilitate fetching the token from an external service. Code change for this is simple.
- Adding test cases to cover this addition.
- Please exclude package.json and the lock from your commit.

Let me know if I can assist in any of the above. :+1:

**Comment by vigie** [comment]
Thanks @kum-deepak , I will certainly add the test case and exclude the pacakge and lock.

>Allow the callback to be optionally async - this will facilitate fetching the token from an external service. Code change for this is simple.

I'm not completely opposed to this but it will introduce asymmetry. At a minimum I feel we would need to make `unsubHeaders` optionally async as well....but why stop there: by the same reasoning `connectHeaders`, `disconnectHeaders` and `websocketFactory` (yes - I need to send my auth-token in the protocol) should all be allowed to become async. At that point `beforeConnect` probably becomes redundant. But I am NOT recommending all of this - I think this would be a fairly clumsy API. Much nicer would be to introduce the interceptors pattern [as I suggest in the issue](https://github.com/stomp-js/rx-stomp/issues/298#issuecomment-797172895), which would centralize this logic. That is a bigger change and unfortunately I have been unable to find the time away from work to do it as yet. 

In the meantime, this little PR makes the current client functionally complete for auth use cases and is consistent with the current API. The workaround I use for the lack of async everywhere/interceptor is to fire off a different process that continually keeps my token refreshed such that when the sync methods are used to retrieve the token, it is guaranteed valid. But that workaround does require this PR to be complete.

To summarize: my recommendation is that we DO NOT introduce the asymmetry of allowing `subscribeHeaders` to be async. Furthermore, we do not in general solve this problem by introducing async callbacks everywhere, rather we go after the interceptors pattern. Interceptor pattern would make all async callbacks redundant, but if in the meantime we introduce them, we'll have to live with them - maintain them and _explain_ them in relation to interceptors (the explanation being "you can choose one or the other approach"). You've done a great job keeping the RxStomp API simple and minimal so far, I'd hate to see it grow in unnecessary and inelegant ways.

If you feel strongly, I will agree to making both `subHeaders` and `unsubHeaders` optionally async in this PR.

**Comment by kum-deepak** [comment]
I think it makes sense. Let us keep it simple. 

Thanks for adding the test cases. I will merge it in a day or two.

**Comment by vigie** [comment]
Great thanks. If you could publish a new version of ng2-stompjs I'd really appreciate it - that's how I consume this.

**Comment by kum-deepak** [comment]
Published 1.1.0 to npm. I usually do not update ng2-stompjs for minor releases of rx-stomp. I uninstall and install `@stomp/ng2-stompjs` which seems to do the trick (I hope to someday figure out an easier way). Please let me know if you would still like me to release an updated ng2-stompjs.
